doc/design.md

```diff
@@ -0,0 +1,316 @@
+# GoTorch
+
+This document explains the motivations and key design challenges of GoTorch.
+
+## GoTorch and Go+
+
+GoTorch includes a Go binding of the C++ core of PyTorch, known as libtorch.
+There are many language bindings of libtorch, including Rust and
+Haskell. However, according to our survey, most Python users don’t feel
+programming in Rust, Haskell, or Julia, is more efficient than in Python.  So,
+language binding does not make much sense alone.
+
+The complete story of GoTorch includes Go+, a language whose syntax is as
+concise as Python, but its compiler generates Go programs. Programming deep
+learning systems in Go+ is hopefully as efficiently as in Python, and Go+
+translates the work into Go source code, which compiles into native code running
+on servers and mobile devices, including phones, pads, and self-driving cars.
+
+In addition to the Go binding of libtorch, GoTorch also includes the other two
+layers of functionalities of PyTorch provided in Python -- torch.nn.functional,
+and torch.nn.
+
+## Layers of Functionalities
+
+Generally, PyTorch provides three layers of APIs, not all of which are in
+libtorch.
+
+1. The finest-grained layer is in libtorch -- about 1600 native functions, each
+   is a fundamental operation in mathematics or its corresponding gradient
+   operation.  Each native function has CPU and GPU implementations. By linking
+   libtorch with github.com/pytorch/xla, we get an additional implementation for
+   Google TPU.  For the layer
+
+1. A higher-level abstraction is in the Python package torch.nn.functional,
+   which provides functions defined in Python and calls native functions in
+   C/C++.
+
+1. The highest layer is modules; each is a Python class with a method forward
+   defining the forward computation process and data members that can store
+   states.
+
+## Tensors and Garbage Collection
+
+libtorch includes the C++ definition of the fundamental data type `at::Tensor`
+and native functions that operate it.
+
+The key design feature of the tensor is automatic garbage collection (GC). In
+C++, the class `at::Tensor` contains only one data member, a `shared_ptr` to the
+implementation of the tensor type.  The smart pointer `shared_ptr` implements
+reference count-based GC, which frees a tensor once its reference count becomes
+zero.  Comparing with Go and Java’s GC, which runs the mark-and-sweep algorithm,
+reference count reacts instantly but cannot handle the case of
+cyclic-dependency.
```

Comment:
`std::weak_ptr` can resolve cyclic-dependency of `std::shared_ptr`s with careful design. `c10::weak_intrusive_ptr` is the intrusive version of `std::weak_ptr`.